Sources/DeepLearning/Random.swift

```diff
@@ -178,11 +169,138 @@ public struct ThreefryRandomNumberGenerator: SeedableRandomNumberGenerator {
+        let lo0 = UInt32(truncatingIfNeeded: (product0 & 0x0000_0000_FFFF_FFFF))
+
+        let product1: UInt64 = roundConstant1 &* UInt64(ctr.2)
+        let hi1 = UInt32(truncatingIfNeeded: product1 >> 32)
+        let lo1 = UInt32(truncatingIfNeeded: (product1 & 0x0000_0000_FFFF_FFFF))
+
+        return (hi1 ^ ctr.1 ^ key.0, lo1, hi0 ^ ctr.3 ^ key.1, lo0)
+    }
+
+    private func calculateRandom(ctr initialCtr: Vector4, key initialKey: Vector2) -> Vector4 {
+        var ctr = initialCtr
+        var key = initialKey
+        // 10 rounds
+        // R1
+        ctr = round(ctr: ctr, key: key)
+        // R2
+        key = bump(key: key)
+        ctr = round(ctr: ctr, key: key)
+        // R3
+        key = bump(key: key)
+        ctr = round(ctr: ctr, key: key)
+        // R4
+        key = bump(key: key)
+        ctr = round(ctr: ctr, key: key)
+        // R5
+        key = bump(key: key)
+        ctr = round(ctr: ctr, key: key)
+        // R6
+        key = bump(key: key)
+        ctr = round(ctr: ctr, key: key)
+        // R7
+        key = bump(key: key)
+        ctr = round(ctr: ctr, key: key)
+        // R8
+        key = bump(key: key)
+        ctr = round(ctr: ctr, key: key)
+        // R9
+        key = bump(key: key)
+        ctr = round(ctr: ctr, key: key)
+        // R10
+        key = bump(key: key)
+        ctr = round(ctr: ctr, key: key)
+
+        return ctr
+    }
+
+    internal init(uint64Seed seed: UInt64) {
+        key = seed.asVector2
+    }
+
+    public init(seed: [UInt8]) {
+        precondition(seed.count > 0, "Length of seed must be positive")
+        precondition(seed.count <= 8, "Length of seed must be at most 8")
+        var combinedSeed: UInt64 = 0
+        for i in 0..<seed.count {
+            combinedSeed += UInt64(seed[i]) << UInt64(8 * i)
+        }
+        self.init(uint64Seed: combinedSeed)
+    }
+
+    public mutating func next() -> UInt64 {
+        if useNextValue {
+            useNextValue = false
+            return nextValue
+        } else {
+            let (this, next) = makeUInt64Pair(
+                from: calculateRandom(ctr: ctr.asVector4, key: key)
             )
-        )
+            useNextValue = true
+            nextValue = next
+            ctr += 1
+            return this
+        }
+    }
+}
+
+/// Private helpers.
+fileprivate extension UInt64 {
+    var asVector2: Vector2 {
+        let msb = UInt32(truncatingIfNeeded: self >> 32)
+        let lsb = UInt32(truncatingIfNeeded: self & 0x0000_0000_FFFF_FFFF)
+        return (msb, lsb)
+    }
+
+    var asVector4: Vector4 {
+        let msb = UInt32(truncatingIfNeeded: self >> 32)
+        let lsb = UInt32(truncatingIfNeeded: self & 0x0000_0000_FFFF_FFFF)
+        return (0, 0, msb, lsb)
     }
+
+    init(fromVector vector: Vector2) {
+        self = (UInt64(vector.0) << 32) + UInt64(vector.1)
+    }
+}
+
+private func makeUInt64Pair(from vector: Vector4) -> (UInt64, UInt64) {
```

Comment:
I think the argument label can be omitted since the function name is a bit factory-ish.